### ISSUE TITLE
Enhancement: Implement ProgressBar controls.

### ISSUE THREAD

**Issue by xystra** [TheCoder4eu/BootsFaces-OSP#245]
Have you researched the idea of implementing progress bars within BootsFaces to programatically support the features implemented by the themes you have added recently (wrapbootstrap)?

**Comment by stephanrauh** [comment]
I've been thinking about progress bars every once in a while. A good progress bar has to communicate with the server in order to show the precise progress (e.g. 65%). Currently, I don't know how to do it. Do you have an idea?

**Comment by xystra** [comment]
Auto-updating the aria-valuenow and width requires scripting to work, and in theory you can show that in your example page, but to just get the progress bar to display is as simple as interpreting

```html
<div class="progress progress-striped active">
    <div class="progress-bar progress-bar-info" role="progressbar" aria-valuenow="20" aria-valuemin="0" aria-valuemax="100" style="width: 20%">
        <span class="sr-only"></span>
    </div>
</div>
```

updating the value is as simple as:
```javascript
var newValue = $('.progress-bar').attr('aria-valuenow') + 20;
$('.progress-bar').css('width', newValue+ '%').attr('aria-valuenow', newValue);
```
but that should be at the users discrepancy. You should just provide the tools to get the progress bar to show up.

**Comment by xystra** [comment]
Just a thought, a user can bind the progress bar to a backing bean, like in dataTables, which implements a bootsfaces interface that can update the values using the UIComponent/UIView:

```Java
final UIViewRoot viewRoot = FacesContext.getCurrentInstance().getViewRoot();
final UIComponent component = viewRoot.findComponent("someId");

final Map<String, Object> attrs  = component.getAttributes();
attrs.put('aria-valuenow', newValue);
attrs.put('style', 'width: newValue %');
```

Might be worth a try....

**Comment by stephanrauh** [comment]
Thanks for providing the snippets! The remaining question is how to update the progress bar. Most progress bars are used on long-running synchronous backend calls. In order to real real progress numbers, we could call a server side method every second or so. This method has to report the progress of the main request, which still runs. That, in turn, means we're in multithreading hell. I've implemented such an interactive progress bar a couple of years ago, so I know it's possible, but currently I shy away from the risk. We can think about the API nonetheless. What about this?

```xml
<progressbar show-on-ajax="true" update-interval="500ms" value="#{myBean.myProgress}" />
```

**Comment by xystra** [comment]
I'm getting some free time coming up in my project, and I might tackle this and see what I can come up with. In the meantime I would add the additional parameters as well:
```
active="true|false"
type="info|alert|warning|success"
```

**Comment by stephanrauh** [comment]
I'm looking forward to it!

**Comment by zhedar** [comment]
I second @xystra's proposal of only providing the general ui of a ProgressBar, that's how they work normally. You can update them via ajax by binding the value or other attributes to a backing back quite easily, so that's completely up to the user.
I actually use them for showing static values only, which may look like this (featuring PrimeFaces' ProgressBar):
<img width="343" alt="bildschirmfoto 2016-04-07 um 17 45 29" src="https://cloud.githubusercontent.com/assets/6633403/14357430/fae1f20e-fce8-11e5-8682-a3611a705965.png">

Server-sided updates would be a total overkill in this scenario.

I would like to look into this, but I don't won't to pass you on this @xystra, so is this something you're currently working on?
I need to get the LibraryGenerating working again first though :)

**Comment by stephanrauh** [comment]
I'm still looking forward to your solution, no matter who implements it first :).

**Comment by xystra** [comment]
@zhedar If you got the time go for it. I barely have had enough time for normal bodily functions with work and kids. I've been thrust onto 2 parallel projects and haven't had the time to put the framework in place. My mind is ready to explode going back and forth from c++ and java.

**Comment by stephanrauh** [comment]
@xystra I guess it's three projects. Kids are a full-time project, too :).

**Comment by zhedar** [comment]
In the meantime I made some progress on the progressBars. However I need to figure out, why it looks a bit boxy and the css isn't functioning how I would expect it to. I may have forgotten to add some dependencies.

So currently that's how it looks like:
<img width="729" alt="bildschirmfoto 2016-04-08 um 21 51 03" src="https://cloud.githubusercontent.com/assets/6633403/14395825/6556c2cc-fdd4-11e5-85e5-569cd4a2a09f.png">

**Now I got a design decision to make, where I wanted get some feedback:**
Currently you need to add the actual value twice, if you want the ProgressBar to show it as a label, e.g.
`<b:progressBar value="50" label="50%" look="danger"/>`
I think that's a bit clunky, so I thought it would be viable to put the label in there automatically, derived from the actual value. That way you would need to set label="" explicitly, if you don't want this number to be shown. But in my opinion the first use case is more common than the latter. So if we document it, that should be fine.

What do you think?

**Comment by stephanrauh** [comment]
Sounds promising! To answer your question:

- I suggest to avoid the word "label" to avoid confusion. I prefer "caption" (for want of a better word).
- Our input fields already use the attribute `render-label="false"`. So I prefer `render-caption="false"` over `caption=""`.

**Comment by zhedar** [comment]
Thanks for the hint, I will probably change that right away.
I can see `render-caption` working, if we default it to true. That way you can hide it by setting that attribute and don't need to change `render-caption` first, if we add a manually created `caption`.

**Comment by zhedar** [comment]
This was just merged in #345 (hey 100 issues ahead), so I guess we can close this. I just forgot the issue close hook in the commit message :/

**Comment by stephanrauh** [comment]
THX!
